cpp/src/arrow/compute/exec/expression.cc

```diff
@@ -62,6 +62,11 @@ Expression call(std::string function, std::vector<Expression> arguments,
   call.function_name = std::move(function);
   call.arguments = std::move(arguments);
   call.options = std::move(options);
+
+  call.hash = std::hash<std::string>{}(call.function_name);
```

Comment:
Hmmm, the Valgrind failure should be fixed if we move computation of the hash into `Expresion::Expression(Call)`

Reply:
Ah, I had missed the Valgrind failure. My bad.